From 11379c06566effc45d78b0ae3d380724ecf8cbea Mon Sep 17 00:00:00 2001
From: Manvi Bajaj <manvi.bajaj@intel.com>
Date: Mon, 16 Dec 2024 11:07:54 +0000
Subject: [PATCH] Fix static analysis issues for thermal daemon for caas

Below issues will be solved with this change.

+------------------------------------+
|Unchecked return value from library |
|Argument cannot be negative         |
|Resource leak                       |
+------------------------------------+

Signed-off-by: Manvi Bajaj <manvi.bajaj@intel.com>
---
 src/android_main.cpp | 28 +++++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/src/android_main.cpp b/src/android_main.cpp
index 1832a52..cbc78a6 100644
--- a/src/android_main.cpp
+++ b/src/android_main.cpp
@@ -125,8 +125,21 @@ static void daemonize(char *rundir, char *pidfile) {
 	}
 
 	i = open("/dev/null", O_RDWR);
-	dup(i);
-	dup(i);
+	if (i == -1)
+	{
+		thd_log_info("Could not open /dev/null\n");
+		exit(EXIT_FAILURE);
+	}
+	int ret = dup(i);
+	if (ret == -1) {
+		close(i);
+		exit(EXIT_FAILURE);
+	}
+	int ret1 = dup(i);
+	if (ret1 == -1) {
+		close(i);
+		exit(EXIT_FAILURE);
+	}
 	chdir(rundir);
 
 	pid_file_handle = open(pidfile, O_RDWR | O_CREAT, 0600);
@@ -148,6 +161,9 @@ static void daemonize(char *rundir, char *pidfile) {
 	thd_log_info("Thermal PID %d\n", getpid());
 	snprintf(str, sizeof(str), "%d\n", getpid());
 	write(pid_file_handle, str, strlen(str));
+	close(ret);
+	close(ret1);
+	close(i);
 }
 
 static void print_usage(FILE* stream, int exit_code) {
@@ -239,7 +255,13 @@ int main(int argc, char *argv[]) {
 			exit(EXIT_FAILURE);
 		}
 	}
-	mkdir(TDCONFDIR, 0755); // Don't care return value as directory
+	if (mkdir(TDCONFDIR, 0755) !=0) {
+		if (errno != EEXIST) {
+			fprintf(stderr, "Cannot create '%s': %s\n", TDCONFDIR,
+					strerror(errno));
+			exit(EXIT_FAILURE);
+		}
+	}
 	if (!no_daemon) {
 		daemonize((char *) "/data/vendor/thermal-daemon",
 				(char *) "/data/vendor/thermal-daemon/thermald.pid");
-- 
2.34.1

